// @elastic/eui/test-env/services/format/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "formatAuto", {
  enumerable: true,
  get: function get() {
    return _format_auto.formatAuto;
  }
});
Object.defineProperty(exports, "formatBoolean", {
  enumerable: true,
  get: function get() {
    return _format_boolean.formatBoolean;
  }
});
Object.defineProperty(exports, "formatDate", {
  enumerable: true,
  get: function get() {
    return _format_date.formatDate;
  }
});
Object.defineProperty(exports, "dateFormatAliases", {
  enumerable: true,
  get: function get() {
    return _format_date.dateFormatAliases;
  }
});
Object.defineProperty(exports, "formatNumber", {
  enumerable: true,
  get: function get() {
    return _format_number.formatNumber;
  }
});
Object.defineProperty(exports, "formatText", {
  enumerable: true,
  get: function get() {
    return _format_text.formatText;
  }
});

var _format_auto = require("./format_auto");

var _format_boolean = require("./format_boolean");

var _format_date = require("./format_date");

var _format_number = require("./format_number");

var _format_text = require("./format_text");
