// @elastic/eui/test-env/components/page/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "EuiPage", {
  enumerable: true,
  get: function get() {
    return _page.EuiPage;
  }
});
Object.defineProperty(exports, "EuiPageBody", {
  enumerable: true,
  get: function get() {
    return _page_body.EuiPageBody;
  }
});
Object.defineProperty(exports, "EuiPageContent", {
  enumerable: true,
  get: function get() {
    return _page_content.EuiPageContent;
  }
});
Object.defineProperty(exports, "EuiPageContentBody", {
  enumerable: true,
  get: function get() {
    return _page_content.EuiPageContentBody;
  }
});
Object.defineProperty(exports, "EuiPageContentHeader", {
  enumerable: true,
  get: function get() {
    return _page_content.EuiPageContentHeader;
  }
});
Object.defineProperty(exports, "EuiPageContentHeaderSection", {
  enumerable: true,
  get: function get() {
    return _page_content.EuiPageContentHeaderSection;
  }
});
Object.defineProperty(exports, "EuiPageHeader", {
  enumerable: true,
  get: function get() {
    return _page_header.EuiPageHeader;
  }
});
Object.defineProperty(exports, "EuiPageHeaderContent", {
  enumerable: true,
  get: function get() {
    return _page_header.EuiPageHeaderContent;
  }
});
Object.defineProperty(exports, "EuiPageHeaderSection", {
  enumerable: true,
  get: function get() {
    return _page_header.EuiPageHeaderSection;
  }
});
Object.defineProperty(exports, "EuiPageSideBar", {
  enumerable: true,
  get: function get() {
    return _page_side_bar.EuiPageSideBar;
  }
});
Object.defineProperty(exports, "EuiPageTemplate", {
  enumerable: true,
  get: function get() {
    return _page_template.EuiPageTemplate;
  }
});

var _page = require("./page");

var _page_body = require("./page_body");

var _page_content = require("./page_content");

var _page_header = require("./page_header");

var _page_side_bar = require("./page_side_bar");

var _page_template = require("./page_template");
