// @elastic/eui/test-env/themes/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "EUI_THEMES", {
  enumerable: true,
  get: function get() {
    return _themes.EUI_THEMES;
  }
});
Object.defineProperty(exports, "isLegacyTheme", {
  enumerable: true,
  get: function get() {
    return _themes.isLegacyTheme;
  }
});
Object.defineProperty(exports, "isDefaultTheme", {
  enumerable: true,
  get: function get() {
    return _themes.isDefaultTheme;
  }
});
Object.defineProperty(exports, "AMSTERDAM_NAME_KEY", {
  enumerable: true,
  get: function get() {
    return _theme.AMSTERDAM_NAME_KEY;
  }
});
Object.defineProperty(exports, "EuiThemeAmsterdam", {
  enumerable: true,
  get: function get() {
    return _theme.EuiThemeAmsterdam;
  }
});
Object.defineProperty(exports, "LEGACY_NAME_KEY", {
  enumerable: true,
  get: function get() {
    return _theme2.LEGACY_NAME_KEY;
  }
});
Object.defineProperty(exports, "EuiThemeLegacy", {
  enumerable: true,
  get: function get() {
    return _theme2.EuiThemeLegacy;
  }
});

var _themes = require("./themes");

var _theme = require("./amsterdam/theme");

var _theme2 = require("./legacy/theme");
