// @elastic/eui/test-env/services/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  keys: true,
  accessibleClickKeys: true,
  cascadingMenuKeys: true,
  comboBoxKeys: true,
  htmlIdGenerator: true,
  useGeneratedHtmlId: true,
  LEFT_ALIGNMENT: true,
  RIGHT_ALIGNMENT: true,
  CENTER_ALIGNMENT: true,
  BREAKPOINTS: true,
  BREAKPOINT_KEYS: true,
  getBreakpoint: true,
  isWithinBreakpoints: true,
  isWithinMaxBreakpoint: true,
  isWithinMinBreakpoint: true,
  isColorDark: true,
  isValidHex: true,
  calculateContrast: true,
  calculateLuminance: true,
  hexToHsv: true,
  hexToRgb: true,
  hsvToHex: true,
  hsvToRgb: true,
  rgbToHex: true,
  rgbToHsv: true,
  VISUALIZATION_COLORS: true,
  DEFAULT_VISUALIZATION_COLOR: true,
  colorPalette: true,
  euiPaletteForLightBackground: true,
  euiPaletteForDarkBackground: true,
  euiPaletteColorBlind: true,
  euiPaletteColorBlindBehindText: true,
  euiPaletteForStatus: true,
  euiPaletteForTemperature: true,
  euiPaletteComplimentary: true,
  euiPaletteNegative: true,
  euiPalettePositive: true,
  euiPaletteCool: true,
  euiPaletteWarm: true,
  euiPaletteGray: true,
  getSteppedGradient: true,
  transparentize: true,
  tint: true,
  shade: true,
  saturate: true,
  desaturate: true,
  lightness: true,
  makeHighContrastColor: true,
  makeDisabledContrastColor: true,
  useColorPickerState: true,
  useColorStopsState: true,
  copyToClipboard: true,
  formatAuto: true,
  formatBoolean: true,
  formatDate: true,
  formatNumber: true,
  formatText: true,
  dateFormatAliases: true,
  isEvenlyDivisibleBy: true,
  isWithinRange: true,
  Pager: true,
  Random: true,
  getSecureRelForTarget: true,
  toSentenceCase: true,
  toInitials: true,
  slugify: true,
  PropertySortType: true,
  SortDirectionType: true,
  SortDirection: true,
  SortableProperties: true,
  Comparators: true,
  calculatePopoverPosition: true,
  findPopoverPosition: true,
  getDurationAndPerformOnFrame: true,
  getTransitionTimings: true,
  getWaitDuration: true,
  performOnFrame: true,
  EuiWindowEvent: true,
  useCombinedRefs: true,
  useForceRender: true,
  useUpdateEffect: true,
  useDependentState: true,
  useIsWithinBreakpoints: true,
  useMouseMove: true,
  isMouseEvent: true,
  throttle: true
};
Object.defineProperty(exports, "accessibleClickKeys", {
  enumerable: true,
  get: function get() {
    return _accessibility.accessibleClickKeys;
  }
});
Object.defineProperty(exports, "cascadingMenuKeys", {
  enumerable: true,
  get: function get() {
    return _accessibility.cascadingMenuKeys;
  }
});
Object.defineProperty(exports, "comboBoxKeys", {
  enumerable: true,
  get: function get() {
    return _accessibility.comboBoxKeys;
  }
});
Object.defineProperty(exports, "htmlIdGenerator", {
  enumerable: true,
  get: function get() {
    return _accessibility.htmlIdGenerator;
  }
});
Object.defineProperty(exports, "useGeneratedHtmlId", {
  enumerable: true,
  get: function get() {
    return _accessibility.useGeneratedHtmlId;
  }
});
Object.defineProperty(exports, "LEFT_ALIGNMENT", {
  enumerable: true,
  get: function get() {
    return _alignment.LEFT_ALIGNMENT;
  }
});
Object.defineProperty(exports, "RIGHT_ALIGNMENT", {
  enumerable: true,
  get: function get() {
    return _alignment.RIGHT_ALIGNMENT;
  }
});
Object.defineProperty(exports, "CENTER_ALIGNMENT", {
  enumerable: true,
  get: function get() {
    return _alignment.CENTER_ALIGNMENT;
  }
});
Object.defineProperty(exports, "BREAKPOINTS", {
  enumerable: true,
  get: function get() {
    return _breakpoint.BREAKPOINTS;
  }
});
Object.defineProperty(exports, "BREAKPOINT_KEYS", {
  enumerable: true,
  get: function get() {
    return _breakpoint.BREAKPOINT_KEYS;
  }
});
Object.defineProperty(exports, "getBreakpoint", {
  enumerable: true,
  get: function get() {
    return _breakpoint.getBreakpoint;
  }
});
Object.defineProperty(exports, "isWithinBreakpoints", {
  enumerable: true,
  get: function get() {
    return _breakpoint.isWithinBreakpoints;
  }
});
Object.defineProperty(exports, "isWithinMaxBreakpoint", {
  enumerable: true,
  get: function get() {
    return _breakpoint.isWithinMaxBreakpoint;
  }
});
Object.defineProperty(exports, "isWithinMinBreakpoint", {
  enumerable: true,
  get: function get() {
    return _breakpoint.isWithinMinBreakpoint;
  }
});
Object.defineProperty(exports, "isColorDark", {
  enumerable: true,
  get: function get() {
    return _color.isColorDark;
  }
});
Object.defineProperty(exports, "isValidHex", {
  enumerable: true,
  get: function get() {
    return _color.isValidHex;
  }
});
Object.defineProperty(exports, "calculateContrast", {
  enumerable: true,
  get: function get() {
    return _color.calculateContrast;
  }
});
Object.defineProperty(exports, "calculateLuminance", {
  enumerable: true,
  get: function get() {
    return _color.calculateLuminance;
  }
});
Object.defineProperty(exports, "hexToHsv", {
  enumerable: true,
  get: function get() {
    return _color.hexToHsv;
  }
});
Object.defineProperty(exports, "hexToRgb", {
  enumerable: true,
  get: function get() {
    return _color.hexToRgb;
  }
});
Object.defineProperty(exports, "hsvToHex", {
  enumerable: true,
  get: function get() {
    return _color.hsvToHex;
  }
});
Object.defineProperty(exports, "hsvToRgb", {
  enumerable: true,
  get: function get() {
    return _color.hsvToRgb;
  }
});
Object.defineProperty(exports, "rgbToHex", {
  enumerable: true,
  get: function get() {
    return _color.rgbToHex;
  }
});
Object.defineProperty(exports, "rgbToHsv", {
  enumerable: true,
  get: function get() {
    return _color.rgbToHsv;
  }
});
Object.defineProperty(exports, "VISUALIZATION_COLORS", {
  enumerable: true,
  get: function get() {
    return _color.VISUALIZATION_COLORS;
  }
});
Object.defineProperty(exports, "DEFAULT_VISUALIZATION_COLOR", {
  enumerable: true,
  get: function get() {
    return _color.DEFAULT_VISUALIZATION_COLOR;
  }
});
Object.defineProperty(exports, "colorPalette", {
  enumerable: true,
  get: function get() {
    return _color.colorPalette;
  }
});
Object.defineProperty(exports, "euiPaletteForLightBackground", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteForLightBackground;
  }
});
Object.defineProperty(exports, "euiPaletteForDarkBackground", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteForDarkBackground;
  }
});
Object.defineProperty(exports, "euiPaletteColorBlind", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteColorBlind;
  }
});
Object.defineProperty(exports, "euiPaletteColorBlindBehindText", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteColorBlindBehindText;
  }
});
Object.defineProperty(exports, "euiPaletteForStatus", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteForStatus;
  }
});
Object.defineProperty(exports, "euiPaletteForTemperature", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteForTemperature;
  }
});
Object.defineProperty(exports, "euiPaletteComplimentary", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteComplimentary;
  }
});
Object.defineProperty(exports, "euiPaletteNegative", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteNegative;
  }
});
Object.defineProperty(exports, "euiPalettePositive", {
  enumerable: true,
  get: function get() {
    return _color.euiPalettePositive;
  }
});
Object.defineProperty(exports, "euiPaletteCool", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteCool;
  }
});
Object.defineProperty(exports, "euiPaletteWarm", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteWarm;
  }
});
Object.defineProperty(exports, "euiPaletteGray", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteGray;
  }
});
Object.defineProperty(exports, "getSteppedGradient", {
  enumerable: true,
  get: function get() {
    return _color.getSteppedGradient;
  }
});
Object.defineProperty(exports, "transparentize", {
  enumerable: true,
  get: function get() {
    return _color.transparentize;
  }
});
Object.defineProperty(exports, "tint", {
  enumerable: true,
  get: function get() {
    return _color.tint;
  }
});
Object.defineProperty(exports, "shade", {
  enumerable: true,
  get: function get() {
    return _color.shade;
  }
});
Object.defineProperty(exports, "saturate", {
  enumerable: true,
  get: function get() {
    return _color.saturate;
  }
});
Object.defineProperty(exports, "desaturate", {
  enumerable: true,
  get: function get() {
    return _color.desaturate;
  }
});
Object.defineProperty(exports, "lightness", {
  enumerable: true,
  get: function get() {
    return _color.lightness;
  }
});
Object.defineProperty(exports, "makeHighContrastColor", {
  enumerable: true,
  get: function get() {
    return _color.makeHighContrastColor;
  }
});
Object.defineProperty(exports, "makeDisabledContrastColor", {
  enumerable: true,
  get: function get() {
    return _color.makeDisabledContrastColor;
  }
});
Object.defineProperty(exports, "useColorPickerState", {
  enumerable: true,
  get: function get() {
    return _color_picker.useColorPickerState;
  }
});
Object.defineProperty(exports, "useColorStopsState", {
  enumerable: true,
  get: function get() {
    return _color_picker.useColorStopsState;
  }
});
Object.defineProperty(exports, "copyToClipboard", {
  enumerable: true,
  get: function get() {
    return _copy_to_clipboard.copyToClipboard;
  }
});
Object.defineProperty(exports, "formatAuto", {
  enumerable: true,
  get: function get() {
    return _format.formatAuto;
  }
});
Object.defineProperty(exports, "formatBoolean", {
  enumerable: true,
  get: function get() {
    return _format.formatBoolean;
  }
});
Object.defineProperty(exports, "formatDate", {
  enumerable: true,
  get: function get() {
    return _format.formatDate;
  }
});
Object.defineProperty(exports, "formatNumber", {
  enumerable: true,
  get: function get() {
    return _format.formatNumber;
  }
});
Object.defineProperty(exports, "formatText", {
  enumerable: true,
  get: function get() {
    return _format.formatText;
  }
});
Object.defineProperty(exports, "dateFormatAliases", {
  enumerable: true,
  get: function get() {
    return _format.dateFormatAliases;
  }
});
Object.defineProperty(exports, "isEvenlyDivisibleBy", {
  enumerable: true,
  get: function get() {
    return _number.isEvenlyDivisibleBy;
  }
});
Object.defineProperty(exports, "isWithinRange", {
  enumerable: true,
  get: function get() {
    return _number.isWithinRange;
  }
});
Object.defineProperty(exports, "Pager", {
  enumerable: true,
  get: function get() {
    return _paging.Pager;
  }
});
Object.defineProperty(exports, "Random", {
  enumerable: true,
  get: function get() {
    return _random.Random;
  }
});
Object.defineProperty(exports, "getSecureRelForTarget", {
  enumerable: true,
  get: function get() {
    return _security.getSecureRelForTarget;
  }
});
Object.defineProperty(exports, "toSentenceCase", {
  enumerable: true,
  get: function get() {
    return _string.toSentenceCase;
  }
});
Object.defineProperty(exports, "toInitials", {
  enumerable: true,
  get: function get() {
    return _string.toInitials;
  }
});
Object.defineProperty(exports, "slugify", {
  enumerable: true,
  get: function get() {
    return _string.slugify;
  }
});
Object.defineProperty(exports, "PropertySortType", {
  enumerable: true,
  get: function get() {
    return _sort.PropertySortType;
  }
});
Object.defineProperty(exports, "SortDirectionType", {
  enumerable: true,
  get: function get() {
    return _sort.SortDirectionType;
  }
});
Object.defineProperty(exports, "SortDirection", {
  enumerable: true,
  get: function get() {
    return _sort.SortDirection;
  }
});
Object.defineProperty(exports, "SortableProperties", {
  enumerable: true,
  get: function get() {
    return _sort.SortableProperties;
  }
});
Object.defineProperty(exports, "Comparators", {
  enumerable: true,
  get: function get() {
    return _sort.Comparators;
  }
});
Object.defineProperty(exports, "calculatePopoverPosition", {
  enumerable: true,
  get: function get() {
    return _popover.calculatePopoverPosition;
  }
});
Object.defineProperty(exports, "findPopoverPosition", {
  enumerable: true,
  get: function get() {
    return _popover.findPopoverPosition;
  }
});
Object.defineProperty(exports, "getDurationAndPerformOnFrame", {
  enumerable: true,
  get: function get() {
    return _transition.getDurationAndPerformOnFrame;
  }
});
Object.defineProperty(exports, "getTransitionTimings", {
  enumerable: true,
  get: function get() {
    return _transition.getTransitionTimings;
  }
});
Object.defineProperty(exports, "getWaitDuration", {
  enumerable: true,
  get: function get() {
    return _transition.getWaitDuration;
  }
});
Object.defineProperty(exports, "performOnFrame", {
  enumerable: true,
  get: function get() {
    return _transition.performOnFrame;
  }
});
Object.defineProperty(exports, "EuiWindowEvent", {
  enumerable: true,
  get: function get() {
    return _window_event.EuiWindowEvent;
  }
});
Object.defineProperty(exports, "useCombinedRefs", {
  enumerable: true,
  get: function get() {
    return _hooks.useCombinedRefs;
  }
});
Object.defineProperty(exports, "useForceRender", {
  enumerable: true,
  get: function get() {
    return _hooks.useForceRender;
  }
});
Object.defineProperty(exports, "useUpdateEffect", {
  enumerable: true,
  get: function get() {
    return _hooks.useUpdateEffect;
  }
});
Object.defineProperty(exports, "useDependentState", {
  enumerable: true,
  get: function get() {
    return _hooks.useDependentState;
  }
});
Object.defineProperty(exports, "useIsWithinBreakpoints", {
  enumerable: true,
  get: function get() {
    return _hooks.useIsWithinBreakpoints;
  }
});
Object.defineProperty(exports, "useMouseMove", {
  enumerable: true,
  get: function get() {
    return _hooks.useMouseMove;
  }
});
Object.defineProperty(exports, "isMouseEvent", {
  enumerable: true,
  get: function get() {
    return _hooks.isMouseEvent;
  }
});
Object.defineProperty(exports, "throttle", {
  enumerable: true,
  get: function get() {
    return _throttle.throttle;
  }
});
exports.keys = void 0;

var keys = _interopRequireWildcard(require("./keys"));

exports.keys = keys;

var _accessibility = require("./accessibility");

var _alignment = require("./alignment");

var _breakpoint = require("./breakpoint");

var _color = require("./color");

var _color_picker = require("./color_picker");

var _console = require("./console");

Object.keys(_console).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _console[key];
    }
  });
});

var _copy_to_clipboard = require("./copy_to_clipboard");

var _format = require("./format");

var _number = require("./number");

var _paging = require("./paging");

var _random = require("./random");

var _security = require("./security");

var _string = require("./string");

var _sort = require("./sort");

var _popover = require("./popover");

var _transition = require("./transition");

var _window_event = require("./window_event");

var _hooks = require("./hooks");

var _throttle = require("./throttle");

var _theme = require("./theme");

Object.keys(_theme).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _theme[key];
    }
  });
});
