// @elastic/eui/test-env/components/modal/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "EuiConfirmModal", {
  enumerable: true,
  get: function get() {
    return _confirm_modal.EuiConfirmModal;
  }
});
Object.defineProperty(exports, "EUI_MODAL_CONFIRM_BUTTON", {
  enumerable: true,
  get: function get() {
    return _confirm_modal.CONFIRM_BUTTON;
  }
});
Object.defineProperty(exports, "EUI_MODAL_CANCEL_BUTTON", {
  enumerable: true,
  get: function get() {
    return _confirm_modal.CANCEL_BUTTON;
  }
});
Object.defineProperty(exports, "EuiModal", {
  enumerable: true,
  get: function get() {
    return _modal.EuiModal;
  }
});
Object.defineProperty(exports, "EuiModalFooter", {
  enumerable: true,
  get: function get() {
    return _modal_footer.EuiModalFooter;
  }
});
Object.defineProperty(exports, "EuiModalHeader", {
  enumerable: true,
  get: function get() {
    return _modal_header.EuiModalHeader;
  }
});
Object.defineProperty(exports, "EuiModalBody", {
  enumerable: true,
  get: function get() {
    return _modal_body.EuiModalBody;
  }
});
Object.defineProperty(exports, "EuiModalHeaderTitle", {
  enumerable: true,
  get: function get() {
    return _modal_header_title.EuiModalHeaderTitle;
  }
});

var _confirm_modal = require("./confirm_modal");

var _modal = require("./modal");

var _modal_footer = require("./modal_footer");

var _modal_header = require("./modal_header");

var _modal_body = require("./modal_body");

var _modal_header_title = require("./modal_header_title");
